recipes/libcusparse/meta.yaml

```
@@ -0,0 +1,159 @@
+      - targets/{{ target_name }}/include             # [linux]
+      - targets/{{ target_name }}/lib/libcusparse.so  # [linux]
+      - targets/{{ target_name }}/lib/stubs           # [linux]
+      # Not shipping Fortran bindings (in src) on Windows for now
+      - targets/{{ target_name }}/libcusparse/src     # [linux]
+      - Library\include                             # [win]
+      - Library\lib                                 # [win]
+    requirements:
+      build:
+        - {{ compiler("c") }}
+        - {{ compiler("cxx") }}
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+        - sysroot_{{ target_platform }} 2.17    # [linux]
+      host:
+        - cuda-version {{ cuda_version }}
+      run:
+        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
+        - {{ pin_subpackage("libcusparse", exact=True) }}
+      run_constrained:
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+        - libcusparse-static >={{ version }}  # [linux]
+    test:
+      commands:
+        - test -L $PREFIX/lib/libcusparse.so                                                   # [linux]
+        - test -f $PREFIX/lib/pkgconfig/cusparse*.pc                                           # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/include/cusparse.h                         # [linux]
+        - test -L $PREFIX/targets/{{ target_name }}/lib/libcusparse.so                         # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/lib/stubs/libcusparse.so                   # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/libcusparse/src/cusparse_fortran.c         # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/libcusparse/src/cusparse_fortran.h         # [linux]
+        - test -f $PREFIX/targets/{{ target_name }}/libcusparse/src/cusparse_fortran_common.h  # [linux]
+        - if not exist %LIBRARY_INC%\cusparse.h exit 1                                         # [win]
+        - if not exist %LIBRARY_LIB%\cusparse.lib exit 1                                       # [win]
+    about:
+      home: https://developer.nvidia.com/cusparse
+      license_file: LICENSE
+      license: LicenseRef-NVIDIA-End-User-License-Agreement
+      license_url: https://docs.nvidia.com/cuda/eula/index.html
+      summary: CUDA Sparse Matrix Library
+      description: |
+        cuSPARSE - Basic Linear Algebra for Sparse Matrices on NVIDIA GPUs
+      doc_url: https://docs.nvidia.com/cuda/cusparse/index.html
+
+  - name: libcusparse-static
+    build:
+      skip: true  # [win]
+    files:
+      - targets/{{ target_name }}/lib/libcusparse_static.a
+    requirements:
+      build:
+        - {{ compiler("c") }}
+        - {{ compiler("cxx") }}
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+        - sysroot_{{ target_platform }} 2.17    # [linux]
+      host:
+        - cuda-version {{ cuda_version }}
+      run:
+        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
```

@robertmaynard Something like this?
```suggestion
      host:
        - cuda-version {{ cuda_version }}
        - libnvjitlink-static
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - {{ pin_compatible("libnvjitlink-static", max_pin="x.x") }}
```

---

Or should it be this? Is it needed in `host`? (I might be unaware of the reason.)
```suggestion
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - libnvjitlink-static
```

---

I think it only needs to be in run, but needs to have the `pin_compatible`

---

> I think it only needs to be in run, but needs to have the `pin_compatible`

I don't think this makes sense. `pin_compatible` needs to refer to the version used in the build/host section, see my suggestion above.

---

pin_compatible only works if it’s in the host section. That is how it determines compatibility. But shouldn’t compatibility (same CTK minor version) be guaranteed by cuda-version?

---

> But shouldn’t compatibility (same CTK minor version) be guaranteed by cuda-version?

@kkraus14 prefers it to be explicit: https://github.com/conda-forge/staged-recipes/pull/21924#discussion_r1155402875 We can revisit this discussion in the feedstocks.

---

libcusolver uses `pin_compatbile` in the `run` section:
```
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - {{ pin_compatible("libcublas", max_pin="x.x") }}
        - {{ pin_compatible("libnvjitlink", max_pin="x.x") }}
```
https://github.com/conda-forge/staged-recipes/blob/5cf9dd7c5648fe5c9eaa0b4c2162ef6a5fd49190/recipes/libcusolver/meta.yaml#L47

---

@robertmaynard please read the lines right above line 47 🙂

---

If it helps, this is the documentation for `pin_compatible`:
https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#extra-jinja2-functions